### PR TITLE
Use SharpCompress for Unpack

### LINKED ISSUE #38: Restoring pure ZIP support
Pure ZIP(`*.zip`) Support may be dropped by this commit. https://github.com/RupertAvery/PSXPackager/commit/ca18d5d7583c404632c9acc6f3f0db655c90f38f

* https://github.com/RupertAvery/PSXPackager/commit/ca18d5d7583c404632c9acc6f3f0db655c90f38f#diff-72a0d633b88d24339c4a9bde59471f398fb093933854c8172901bf709b934627
* https://github.com/RupertAvery/PSXPackager/commit/ca18d5d7583c404632c9acc6f3f0db655c90f38f#diff-e88afaafeafa4678776f48d256eac77a6aa6b63be87f894a97d2a4414405fbf2

see also. [GUI Batch Page](https://github.com/RupertAvery/PSXPackager/blob/764ac5ee03aae02677eaa50bcf30fb5c43dbd7ec/PSXPackagerGUI/Pages/BatchController.cs)

### LINKED ISSUE #39: Use SharpCompress
SharpCompress is pure C# implementation for archives.

https://github.com/adamhathcock/sharpcompress

If use it and no more need to distribute `7z.dll` and or.

SharpCompress - Simple Decompress Example
* https://github.com/takano32/SharpDecompress

SharpCompress - EventHandler
* https://github.com/adamhathcock/sharpcompress/blob/33cd1f3db801cd7868d34844b0246f5a57a73d0d/src/SharpCompress/Archives/IArchive.cs#L10-L17

### PR DESCRIPTION
**Pull request by takano32**
closes #39, closes #38

[SharpCompress](https://github.com/adamhathcock/sharpcompress) is a compression library in pure C#.

Using this make Unpack available in Linux and macOS.

### CODE REVIEW

**Comment by takano32** [comment]
I am debugging a problem where the percentage values are somehow wrong, but I am not sure.

**Comment by takano32** [comment]
The progress indicator is odd, but it works very well.

After all, it is a very nice experience to be able to convert files that are still compressed on Linux and macOS.

So I really want to get rid of this strange bug, but it may be difficult to get rid of it on my own.

Would you have any idea or knowledge for archiver?

Now I decided to make this PR Ready for Preview, but I’m very sorry this PR having indicator bug.The calculation of percentage may be wrong… 🙈

**Comment by RupertAvery** [comment]
let me take a look at your branch. I need to catch up to your changes :)

**Comment by RupertAvery** [comment]
I'm trying out SharpCompress on my 7z files and I get this error:

`SharpCompress.Common.ArchiveException: 'Failed to locate the Zip Header'`

Would you know anything about it?

**Comment by RupertAvery** [comment]
I could be wrong about CompressedBytesRead. 7zip is weird in that you need to go through all previous files before you extract one file.

I don't see the progress update after extracting the first file. It doesn't do anything for a while, then jumps to 98% and finishes. Same with the last file.

Could be that  CompressedBytesRead is the total extracted bytes.  I'll look into it a bit more.  It's nice to have a managed 7z unpacker though.

**Comment by RupertAvery** [comment]
I'm a bit confused. The property is "CompressedBytesRead", but it matches up more or less with the (uncompressed) Size.

Also the CompressedSize is 0.

At any rate, it's good enough for a progressbar.

**Comment by RupertAvery** [comment]
So I tested the percent calculation on the latest branch of SharpCompress (standalone, not using PSXPacakger) and funny enough, the percent reaches 100 all the time for all files.

Dunno if there are updates to the codebase that fix it, or I'm doing something wrong in PSXPackager (probably not, because I copied our code into a new project in SharpCompress to test it, and also using BoF3.

I also looked into the 7z unpacking part of SharpCompress, and yeah, there really is a delay when you unzip files further into the archive. It looks like there is an ongoing issue with it, as the 7zip format requires you to unpack/read the previous file in the archive before you unpack the one you want.

It's not a huge deal, but if you have many .bins in one archive, it will be much slower unpacking.

What happens is, suppose you have the files:

```
BoF3 (1).bin
BoF3 (2).bin
BoF3.cue
```

It will perform the following:

```
unpack BoF3 (1).bin -> output file
unpack BoF3 (1).bin in memory (throw away data) and count how many bytes have been "extracted", when EOF reached,
unpack BoF3 (2).bin -> output file
unpack BoF3 (1).bin in memory (throw away data) and count how many bytes have been "extracted", when EOF reached,
unpack BoF3 (2).bin in memory (throw away data) and count how many bytes have been "extracted", when EOF reached,
unpack BoF3.cue -> output file
```

This also happens in 7zip when you just want to extract BoF3.cue, but if you extact all the files, it can continue from the last position in the archive, I guess.

**Comment by takano32** [comment]
In fact, I did not expect to be treated so courteously! 
Thank you very much!

I see that my code was trying to calculate the progress of individual files from the size of a single compressed file, and that's why it wasn't working.
I understand that calculating from the size of each individual file works fine.

I didn't expect the 7z expansion to be so complicated. Thank you for explaining the details.
I learned a lot. I am not sure if I understood everything, though. 😅 

Finally, this PR is now truly `Ready for Preview`! 🎉 
Thank you very much. 🙏

**Comment by takano32** [comment]
If there seems to be a concern that using SharpCompress will degrade performance, perhaps it would be better to try using SevenZipSharp first and fall back if it is not available, instead of using SharpCompress only?

About 90% of the files I handle are in ZIP format and maybe 10% in 7z format.
I am hopeful that replacing the library with SharpCompress will allow me to handle ZIP files, even if it will degrade performance.

But it is not my intention to be disparaging about performance.

**Comment by RupertAvery** [comment]
Don't worry about it. I was just mentioning it since I noticed the pause when extracting, and I thought you should know.

Eventually someone will come along and fix it. I tried, but I don't know enough about the LZMA format and the codebase.

I'm sure Linux users will be glad to have 7z support!

File: PSXPackager/Program.cs [hunk 1]
```diff
@@ -173,10 +170,9 @@ static void Main(string[] args)
 
         private static IEnumerable<string> GetFilesFromDirectory(string path, string filterExpression, bool recursive)
         {
+            // https://github.com/adamhathcock/sharpcompress/blob/master/FORMATS.md
             var supportedFiles = new List<string>() {
-#if SEVENZIP
-            ".7z", ".gz", ".rar", ".tar", 
-#endif
+            "rar", "zip", "tar", ".gz", ".7z",
```
Comment (RupertAvery) [inline PSXPackager/Program.cs:175]: .rar, .zip, .tar

Comment (RupertAvery) [reply PSXPackager/Program.cs:175]: I can fix this after I merge it in

**Review by RupertAvery — APPROVED** [review]
One small thing found, but I'll fix it after I merge it in. Thanks @takano32!

File: PSXPackager/Program.cs (hunk 1 above)
Comment (takano32) [reply PSXPackager/Program.cs:175]: Thank you for fixing! https://github.com/RupertAvery/PSXPackager/commit/a08fe50e702e30af9acf3a3ae20b5ce17d7575f2

**Pull request merged**